### PR TITLE
`sage.rings`: Modularization fixes (imports), `# needs`

### PR DESCRIPTION
**Pull request by mkoeppe**
- Cherry-picked from #35095 

### :memo: Checklist

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

File: src/sage/rings/function_field/ideal.py [hunk 1]
```diff
@@ -455,20 +455,20 @@ def divisor(self):
...
-            sage: # needs sage.rings.finite_rings
+            sage: # needs sage.modules sage.rings.finite_rings
             sage: Oinf = K.maximal_order_infinite()
             sage: I = Oinf.ideal((x + 1)/(x^3 + 1))
             sage: I.divisor()
             2*Place (1/x)
 
-            sage: # needs sage.rings.function_field
+            sage: # needs sage.modules sage.rings.function_field
             sage: K.<x> = FunctionField(GF(2)); _.<T> = PolynomialRing(K)
```
Comment (kwankyu) [inline src/sage/rings/function_field/ideal.py]: Doesn't `sage.rings.function_field` depend on `sage.modules`?

Comment (mkoeppe) [reply src/sage/rings/function_field/ideal.py]: I'm not sure right now. This dependency is not declared in sage.features, and there are quite a number of these `# needs sage.modules sage.rings.function_field` annotations already elsewhere

Comment (kwankyu) [reply src/sage/rings/function_field/ideal.py]: `sage.rings.function_field` (the python module) depends on (uses) `sage.modules`. Then the distribution package that provides the feature `sage.rings.function_field` should also provide the feature `sage.modules`. Isn't this the design?

Comment (mkoeppe) [reply src/sage/rings/function_field/ideal.py]: Well, looking at `git grep -E 'sage.(modules|matrix).*import' src/sage/rings/function_field`, I see that most places are careful to do the imports when they are needed in individual methods.

On the other hand, the distiributions per #35095: 
```
git --no-pager grep 'sage_setup:' src/sage/rings/function_field
src/sage/rings/function_field/all.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/all__sagemath_categories.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/all__sagemath_modules.py:# sage_setup: distribution = sagemath-modules
src/sage/rings/function_field/all__sagemath_pari.py:# sage_setup: distribution = sagemath-pari
src/sage/rings/function_field/all__sagemath_singular.py:# sage_setup: distribution = sagemath-singular
src/sage/rings/function_field/all__sagemath_symbolics.py:# sage_setup: distribution = sagemath-symbolics
src/sage/rings/function_field/constructor.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/derivations.py:# sage_setup: distribution = sagemath-modules
src/sage/rings/function_field/derivations_polymod.py:# sage_setup: distribution = sagemath-singular
src/sage/rings/function_field/derivations_rational.py:# sage_setup: distribution = sagemath-modules
src/sage/rings/function_field/differential.py:# sage_setup: distribution = sagemath-modules
src/sage/rings/function_field/divisor.py:# sage_setup: distribution = sagemath-modules
src/sage/rings/function_field/element.pxd:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/element.pyx:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/element_polymod.pyx:# sage_setup: distribution = sagemath-singular
src/sage/rings/function_field/element_rational.pyx:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/extensions.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/function_field.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/function_field_polymod.py:# sage_setup: distribution = sagemath-singular
src/sage/rings/function_field/function_field_rational.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/hermite_form_polynomial.pyx:# sage_setup: distribution = sagemath-modules
src/sage/rings/function_field/ideal.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/ideal_polymod.py:# sage_setup: distribution = sagemath-singular
src/sage/rings/function_field/ideal_rational.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/maps.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/order.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/order_basis.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/order_polymod.py:# sage_setup: distribution = sagemath-singular
src/sage/rings/function_field/order_rational.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/place.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/place_polymod.py:# sage_setup: distribution = sagemath-singular
src/sage/rings/function_field/place_rational.py:# sage_setup: distribution = sagemath-categories
src/sage/rings/function_field/valuation.py:# sage_setup: distribution = sagemath-pari
```
(where **sagemath-singular** does depend on **sagemath-modules**).

So based on this, I agree with you

Comment (mkoeppe) [reply src/sage/rings/function_field/ideal.py]: I've simplified it accordingly in 7a2f540

File: src/sage/categories/finite_dimensional_modules_with_basis.py [hunk 2]
```diff
@@ -354,7 +354,7 @@ def echelon_form(self, elements, row_reduced=False, order=None):
                 sage: R.<x,y> = QQ[]
                 sage: C = CombinatorialFreeModule(R, range(3), prefix='x')              # needs sage.modules
                 sage: x = C.basis()                                                     # needs sage.modules
-                sage: C.echelon_form([x[0] - x[1], 2*x[1] - 2*x[2], x[0] - x[2]])       # needs sage.modules sage.rings.function_field
+                sage: C.echelon_form([x[0] - x[1], 2*x[1] - 2*x[2], x[0] - x[2]])       # needs sage.rings.function_field
```
Comment (kwankyu) [inline src/sage/categories/finite_dimensional_modules_with_basis.py]: Here `# needs sage.modules` is right.

File: src/sage/rings/function_field/element_rational.pyx [hunk 3]
```diff
@@ -74,10 +74,11 @@ cdef class FunctionFieldElement_rational(FunctionFieldElement):
             sage: K.<t> = FunctionField(GF(7))
             sage: t.element()
             t
-            sage: type(t.element())                                                     # needs sage.rings.finite_rings
+            sage: type(t.element())                                                     # needs sage.libs.ntl
             <... 'sage.rings.fraction_field_FpT.FpTElement'>
```
Comment (kwankyu) [inline src/sage/rings/function_field/element_rational.pyx:78]: I wonder why `ntl` is involved. Here.

**Comment by kwankyu** [comment]
Otherwise lgtm.

File: src/sage/rings/function_field/element_rational.pyx (hunk 3 above)
Comment (mkoeppe) [reply src/sage/rings/function_field/element_rational.pyx:78]: fraction_field_FpT is implemented using NTL

File: src/sage/categories/finite_dimensional_modules_with_basis.py (hunk 2 above)
Comment (mkoeppe) [reply src/sage/categories/finite_dimensional_modules_with_basis.py]: Thanks, fixed

File: src/sage/rings/function_field/element_rational.pyx (hunk 3 above)
Comment (kwankyu) [reply src/sage/rings/function_field/element_rational.pyx:78]: OK.

**Review by kwankyu — APPROVED** [review]
Thanks.

**Comment by mkoeppe** [comment]
Thank you!

**Pull request merged**